File: windows/src/Notchle.Core/Contracts/Game.cs
namespace Notchle.Core;

// Mirror of Sources/NotchleCore/Contracts/Game.swift. The game is a pure reducer:
// GameEngine.Send(action) replaces State and returns effects for the platform layer to run.
//
// Agreed rules (Tren, 2026-09-24):
// - Heardle-style tiers: 5s, a wrong guess offers a retry at 10s, then 15s. Wrong at the last
//   tier (or giving up) reveals the answer: that track is missed.
// - A guess needs title and every credited artist, any order. Typos are forgiven as long as
//   the gist is right (FuzzyAnswerJudge; shared vectors in /spec/judge-cases.json).
// - Correct: confetti, and the song keeps playing. Next moves on.
// - Set end (Tren, 2026-09-24: "add 20 new songs, or keep the songs you didn't get correctly"):
//   at every set end, complete or failed, the tracks answered correctly join ClearedTrackIds
//   (persisted). Then StartSet(choice), in SetComplete and SetFailed:
//   Replay: the same tracks reshuffled, results cleared (SetNumber unchanged).
//   KeepMisses: the missed tracks plus new ones (not cleared, not in the finished set) up to
//   SetSize, shuffled. At SetComplete (no misses) it equals AllNew.
//   AllNew: SetSize new tracks; fewer left gives a smaller set; none left gives Exhausted.
//   NextSet = StartSet(AllNew) and ReplaySet = StartSet(Replay) remain as aliases.
// - Snippets start at GameConfig.SnippetStart (default 0).
// - Restart (Tren, 2026-09-24): in PlayingSnippet / Guessing it replays the current snippet from
//   its start at the same tier (no attempt used, nothing recorded). In Correct / Revealed it
//   restarts the whole song from 0:00 and lets it play on. Ignored everywhere else, Wrong
//   included: Retry is the way on from there, and a free replay would allow unlimited guesses
//   at the same tier.
// - Skip (Tren, 2026-09-24: "forfeit 1 chance to get the longer version, over completely
//   forfeiting by giving up"): in PlayingSnippet / Guessing it spends the attempt without a
//   guess and plays the next, longer tier. At the last tier it is GiveUp. Ignored elsewhere.

public abstract record GamePhase
{
    public sealed record Idle : GamePhase;
    public sealed record Loading : GamePhase;
    /// Snippet is playing; guess fields may already be used, submitting stops the snippet.
    public sealed record PlayingSnippet(int TierIndex) : GamePhase;
    public sealed record Guessing(int TierIndex) : GamePhase;
    /// Wrong guess with a longer tier left: offer Retry and Give up.
    public sealed record Wrong(int TierIndex, Verdict Verdict) : GamePhase;
    /// Right: confetti, song keeps playing until Next.
    public sealed record Correct(int TierIndex) : GamePhase;
    /// Missed: answer shown, song keeps playing until Next. Verdict is null after a give-up.
    public sealed record Revealed(Verdict? Verdict) : GamePhase;
    public sealed record SetComplete(int CorrectCount) : GamePhase;
    public sealed record SetFailed(int CorrectCount) : GamePhase;
    public sealed record Exhausted : GamePhase;
    /// Next skips the track (missed); Reset goes idle.
    public sealed record Error(string Message) : GamePhase;
}

/// Immutable snapshot; the engine replaces it with `with` expressions.
public sealed record GameState
{
    public required GameConfig Config { get; init; }
    public GamePhase Phase { get; init; } = new GamePhase.Idle();
    public SourceListing? Listing { get; init; }
    public IReadOnlyList<Track> CurrentSet { get; init; } = Array.Empty<Track>();
    public int Index { get; init; }
    public IReadOnlyList<TrackOutcome> Results { get; init; } = Array.Empty<TrackOutcome>();
    /// 1-based set counter within this listing.
    public int SetNumber { get; init; }
    /// Ids of tracks answered correctly at a set end (any set, complete or failed); persisted so relaunches don't repeat them.
    public IReadOnlySet<string> ClearedTrackIds { get; init; } = new HashSet<string>();
    /// Incremented on every correct guess; the UI fires confetti when it changes.
    public int CelebrationCount { get; init; }

    public Track? CurrentTrack => Index >= 0 && Index < CurrentSet.Count ? CurrentSet[Index] : null;
    public int CorrectCount => Results.Count(r => r is TrackOutcome.Correct);
}

/// How the next set is built at a set end.
public enum SetChoice
{
    /// The same tracks again, reshuffled.
    Replay,
    /// The tracks missed in this set, filled up with new ones.
    KeepMisses,
    /// Only new tracks (not cleared, not in the finished set).
    AllNew,
}

public abstract record GameAction
{
    public sealed record Load(SourceRef Ref) : GameAction;
    public sealed record Loaded(SourceListing Listing) : GameAction;
    public sealed record LoadFailed(string Message) : GameAction;
    public sealed record SnippetFinished : GameAction;
    public sealed record Submit(Guess Guess) : GameAction;
    public sealed record Retry : GameAction;
    public sealed record GiveUp : GameAction;
    /// From PlayingSnippet/Guessing: give up this attempt without guessing and play the next,
    /// longer tier. At the last tier it behaves like GiveUp. Ignored elsewhere (in Wrong, Retry
    /// already does this).
    public sealed record Skip : GameAction;
    public sealed record Next : GameAction;
    /// SetComplete / SetFailed only: start the next set as <paramref name="Choice"/> says.
    public sealed record StartSet(SetChoice Choice) : GameAction;
    /// Alias of StartSet(AllNew).
    public sealed record NextSet : GameAction;
    /// Alias of StartSet(Replay).
    public sealed record ReplaySet : GameAction;
    /// Replay the current snippet from its start (PlayingSnippet/Guessing: doesn't use up an
    /// attempt), or restart the whole song from 0:00 (Correct/Revealed). Ignored elsewhere.
    public sealed record Restart : GameAction;
    public sealed record PlaybackFailed(string Message) : GameAction;
    public sealed record Configure(GameConfig Config) : GameAction;
    public sealed record Reset : GameAction;
}

public abstract record GameEffect
{
    public sealed record Fetch(SourceRef Ref) : GameEffect;
    /// Play Seconds of Track from Start, then pause and report SnippetFinished.
    public sealed record PlaySnippet(Track Track, double Start, double Seconds) : GameEffect;
    /// Cancel a running snippet and let the song play on from where it is.
    public sealed record ContinuePlaying : GameEffect;
    public sealed record Stop : GameEffect;
    /// Cancel any running snippet and play Track from the very start, continuing to the end.
    public sealed record RestartTrack(Track Track) : GameEffect;
    /// ClearedTrackIds changed; persist it.
    public sealed record PersistProgress : GameEffect;
}

public interface IAnswerJudge
{
    Verdict Judge(Guess guess, Track track);
}

File: Sources/NotchleCore/Contracts/Game.swift
import Foundation

// The game is a pure reducer: `GameEngine.send(_:)` takes an action, mutates `GameState`
// and returns effects for the platform layer to run (fetch, play, stop). No timers, no I/O,
// no randomness except the injected seed, so it is fully unit-testable and portable.
//
// Agreed rules (Tren, 2026-09-24):
// - Heardle-style tiers: snippet of 5s, a wrong guess offers a retry at 10s, then 15s.
//   A wrong guess at the last tier (or giving up) reveals the answer: that track is missed.
// - A guess needs both title and artists: every credited artist, in any order. Typos are
//   forgiven as long as the gist is right (Tren, 2026-09-24; see FuzzyAnswerJudge).
// - Correct: confetti, and the song keeps playing. `next` moves on.
// - End of a set (Tren, 2026-09-24: "add 20 new songs, or keep the songs you didn't get
//   correctly"): every track answered correctly in the set is cleared (whether or not the set
//   was 20/20) and persisted. Then `startSet(SetChoice)`, from setComplete or setFailed:
//   `.replay` the same tracks reshuffled; `.keepMisses` the missed tracks plus new ones up to
//   the set size; `.allNew` a set of new tracks (not cleared, not in the set just played).
//   Fewer new tracks left gives a smaller set; none left (and nothing kept) gives `exhausted`,
//   paste a new URL. `nextSet`/`replaySet` are aliases for `.allNew`/`.replay`.
// - Snippets start at the start of the song (`GameConfig.snippetStart`).
// - Restart (Tren, 2026-09-24: "a restart button so i can restart the song"): while guessing
//   (`playingSnippet`/`guessing`) it replays the current tier's snippet from its start and
//   costs no attempt; after `correct`/`revealed` it plays the whole song again from 0:00.
//   Not in `wrong`: Retry is the way on there, and a free replay would allow unlimited
//   guesses at the same tier.
// - Skip (Tren, 2026-09-24: "forfeit 1 chance to get the longer version, over completely
//   forfeiting by giving up"): while guessing, give up this attempt without guessing and hear
//   the next, longer tier. At the last tier it is `giveUp`. Not in `wrong` (Retry is that).

public enum GamePhase: Sendable, Hashable {
    /// Nothing loaded. UI asks for a Spotify URL.
    case idle
    case loading
    /// Snippet is playing. Guess fields may already be used; submitting stops the snippet.
    case playingSnippet(tierIndex: Int)
    /// Snippet finished; waiting for a guess.
    case guessing(tierIndex: Int)
    /// Wrong guess and a longer tier is left: UI offers Retry (and Give up).
    case wrong(tierIndex: Int, verdict: Verdict)
    /// Right. Confetti; the song keeps playing until `next`.
    case correct(tierIndex: Int)
    /// Missed; answer is shown and the song keeps playing until `next`.
    /// `verdict` is the last wrong guess, nil if the player gave up without guessing.
    case revealed(verdict: Verdict?)
    /// All tracks of the set answered and every one correct.
    case setComplete(correctCount: Int)
    /// All tracks of the set answered, at least one missed.
    case setFailed(correctCount: Int)
    /// No unplayed tracks left in the listing.
    case exhausted
    /// Loading or playback failed. `next` skips the track (counts as missed); `reset` goes idle.
    case error(message: String)
}

public struct GameState: Sendable, Hashable {
    public var config: GameConfig
    public var phase: GamePhase = .idle
    public var listing: SourceListing?
    /// The current set, in play order.
    public var currentSet: [Track] = []
    /// Index into `currentSet` of the track being played.
    public var index: Int = 0
    /// One entry per finished track of the current set.
    public var results: [TrackOutcome] = []
    /// 1-based set counter within this listing.
    public var setNumber: Int = 0
    /// Track ids answered correctly at the end of a set, per listing. Persisted by the
    /// platform layer so a relaunch does not repeat songs you already cleared.
    public var clearedTrackIDs: Set<String> = []
    /// Incremented on every correct guess. The UI fires confetti when it changes.
    public var celebrationCount: Int = 0

    public var currentTrack: Track? {
        currentSet.indices.contains(index) ? currentSet[index] : nil
    }

    public var correctCount: Int {
        results.filter { if case .correct = $0 { return true } else { return false } }.count
    }

    public init(config: GameConfig = GameConfig(), clearedTrackIDs: Set<String> = []) {
        self.config = config
        self.clearedTrackIDs = clearedTrackIDs
    }
}

public enum GameAction: Sendable, Hashable {
    /// User pasted a URL. Effect: `.fetch`.
    case load(SourceRef)
    /// Platform finished the fetch.
    case loaded(SourceListing)
    case loadFailed(message: String)
    /// Player reports the snippet has played and paused.
    case snippetFinished
    case submit(Guess)
    /// From `.wrong`: play the next (longer) tier.
    case retry
    /// From `.playingSnippet`, `.guessing` or `.wrong`: reveal, counts as missed.
    case giveUp
    /// From playingSnippet/guessing: give up this attempt without guessing and play the next,
    /// longer tier. At the last tier it behaves like giveUp. Ignored elsewhere (in wrong, Retry
    /// already does this).
    case skip
    /// Replay the current snippet from its start (playingSnippet/guessing: doesn't use up an
    /// attempt), or restart the whole song from 0:00 (correct/revealed). Ignored elsewhere.
    case restart
    /// From `.correct`, `.revealed` or `.error`: move to the next track (or end the set).
    case next
    /// From `.setComplete` or `.setFailed`: start the next set as chosen. Ignored elsewhere.
    case startSet(SetChoice)
    /// Alias for `startSet(.allNew)`.
    case nextSet
    /// Alias for `startSet(.replay)`.
    case replaySet
    case playbackFailed(message: String)
    case configure(GameConfig)
    /// Back to `.idle`, keeping `clearedTrackIDs`.
    case reset
}

/// What to play after a set ends.
public enum SetChoice: String, Sendable, Hashable, Codable, CaseIterable {
    /// The same tracks, reshuffled. Keeps the set number.
    case replay
    /// The missed tracks plus new ones up to the set size, shuffled together.
    case keepMisses
    /// Only new tracks: not cleared and not in the set just played.
    case allNew
}

public enum GameEffect: Sendable, Hashable {
    case fetch(SourceRef)
    /// Play `seconds` of `track` from `start`, then pause and report `.snippetFinished`.
    /// Starting a new snippet cancels any snippet still running.
    case playSnippet(Track, start: Double, seconds: Double)
    /// Cancel a running snippet (if any) and let the current song play on from where it is.
    case continuePlaying
    /// Cancel any running snippet and play `track` from the very start, continuing to the end.
    case restartTrack(Track)
    /// Stop playback entirely.
    case stop
    /// `clearedTrackIDs` changed; persist it.
    case persistProgress
    /// The outcome of `track` was just decided: guessed right, or missed (wrong at the last
    /// tier, give up, skip at the last tier, or `next` from an error before an outcome).
    /// Emitted exactly once per played track, first in that step's effect list. The platform
    /// stamps the date and the listing and appends it to the play history.
    /// `wrongGuesses`/`skips` count this track's wrong guesses and skips (a skip at the last
    /// tier is exactly `giveUp`, so it is not counted); both reset per track and on `replaySet`.
    case recordOutcome(Track, TrackOutcome, wrongGuesses: Int, skips: Int)
}

/// Decides whether a guess matches a track. Implemented by `FuzzyAnswerJudge` (Wave 2, agent C).
public protocol AnswerJudging: Sendable {
    func judge(_ guess: Guess, against track: Track) -> Verdict
}
